packages/cel/src/type.ts

```diff
@@ -12,117 +12,310 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-import type { DescMessage, MessageShape } from "@bufbuild/protobuf";
-import type { CelType, CelVal } from "./value/value.js";
-import type { CelList } from "./list.js";
-import type { CelMap } from "./map.js";
-import type { CelUint } from "./uint.js";
+import type { DescMessage, Message, MessageShape } from "@bufbuild/protobuf";
+import { isCelList, type CelList } from "./list.js";
+import { isCelMap, type CelMap } from "./map.js";
+import { isCelUint, type CelUint } from "./uint.js";
+import { isNullMessage, type NullMessage } from "./null.js";
+import {
+  isReflectMessage,
+  type ReflectList,
+  type ReflectMap,
+  type ReflectMessage,
+} from "@bufbuild/protobuf/reflect";
+import { TimestampSchema, DurationSchema } from "@bufbuild/protobuf/wkt";
+
+const privateSymbol = Symbol.for("@bufbuild/cel/type");
 
 /**
  * Types of CEL values.
  *
  * Ref: https://github.com/google/cel-spec/blob/master/doc/langdef.md#values
  */
-export type CelValueType =
-  | CelScalar
-  | CelListValueType
-  | CelMapValueType
-  | DescMessage;
+export type CelType =
+  | CelListType
+  | CelMapType
+  | CelObjectType
+  | CelTypeType
+  | CelScalarType;
 
 /**
  * Scalar CEL value types.
  */
-export enum CelScalar {
-  INT,
-  UINT,
-  BOOL,
-  STRING,
-  BYTES,
-  DOUBLE,
-  NULL,
-  ANY,
-  TYPE,
-}
+// biome-ignore format: Ternaries
+export const CelScalar = Object.freeze({
```

Comment:
I am not sure if we want to use `Object.freeze` as we don't use it elsewhere, but a global object seemed like a good enough use case.

Reply:
Object.freeze isn't wrong, but for consistency, I'd prefer that we stick with `as const`. (Even TypeScript enumerations transpile to unfrozen JS.)